Sources/GRPC/GRPCKeepaliveHandlers.swift

```diff
@@ -0,0 +1,297 @@
+      break
+    }
+
+    context.fireChannelRead(data)
+  }
+
+  private func perform(action: PingHandler.Action?, context: ChannelHandlerContext) {
+    guard let action = action else { return }
+    switch action {
+    case let .schedulePing(delay, timeout):
+      self.schedulePing(delay: delay, timeout: timeout, context: context)
+    case .cancelScheduledTimeout:
+      self.cancelScheduledTimeout()
+    case let .reply(payload):
+      self.send(payload: payload, context: context)
+    }
+  }
+
+  private func send(payload: HTTP2Frame.FramePayload, context: ChannelHandlerContext) {
+    let frame = self.wrapOutboundOut(.init(streamID: .rootStream, payload: payload))
+    context.writeAndFlush(frame, promise: nil)
+  }
+
+  private func close(context: ChannelHandlerContext) {
+    self.connectionManager?.idle()
+    context.close(mode: .all, promise: nil)
+  }
+
+  private func schedulePing(delay: TimeAmount, timeout: TimeAmount, context: ChannelHandlerContext) {
+    context.eventLoop.scheduleTask(in: delay) { [unowned self] in
+      self.perform(action: self.pingHandler.pingFired(), context: context)
+      self.scheduleClose(timeout: timeout, context: context)
+    }
+  }
+
+  private func scheduleClose(timeout: TimeAmount, context: ChannelHandlerContext) {
+    self.scheduledClose = context.eventLoop.scheduleTask(in: timeout) { [unowned self] in
+      self.close(context: context)
+    }
+  }
+
+  private func cancelScheduledTimeout() {
+    self.scheduledClose?.cancel()
+    self.scheduledClose = nil
+  }
+}
+
+final class PingHandler {
+  /// Code for ping
+  private let pingCode: UInt64
+
+  /// The amount of time to wait before sending a keepalive ping.
+  private let interval: TimeAmount
+
+  /// The amount of time to wait for an acknowledgment.
+  /// If it does not receive an acknowledgment within this time, it will close the connection
+  private let timeout: TimeAmount
+
+  /// Send keepalive pings even if there are no calls in flight.
+  private let permitWithoutCalls: Bool
+
+  /// Maximum number of pings that can be sent when there is no data/header frame to be sent.
+  private let maximumPingsWithoutData: UInt
+
+  /// If there are no data/header frames being received:
+  /// The minimum amount of time to wait between successive pings.
+  private let minimumSentPingIntervalWithoutData: TimeAmount
+
+  /// If there are no data/header frames being sent:
+  /// The minimum amount of time expected between receiving successive pings.
+  /// If the time between successive pings is less than this value, then the ping will be considered a bad ping from the peer.
+  /// Such a ping counts as a "ping strike".
+  private let minimumReceivedPingIntervalWithoutData: TimeAmount?
+
+  /// Maximum number of bad pings that the server will tolerate before sending an HTTP2 GOAWAY frame and closing the connection.
+  /// Setting it to `0` allows the server to accept any number of bad pings.
+  private let maximumPingStrikes: UInt?
```

Comment:
`nil` means we're a client and ping strikes aren't applicable: we should document that here. I'd also be open to having an enum here to codify this.